File: src/lib/api/gateway.ts
import { invoke } from "@tauri-apps/api/core";

export type ProviderProtocol =
  | "openai_compatible"
  | "anthropic_compatible"
  | "codex_experimental";

export interface GatewayStatus {
  running: boolean;
  host: string;
  port: number;
  openai_base_url: string;
  anthropic_base_url: string;
}

export interface ApiKeyStatus {
  exists: boolean;
  preview: string | null;
}

export interface Provider {
  id: string;
  name: string;
  base_url: string;
  protocol: ProviderProtocol;
  enabled: boolean;
  default_model: string;
  models: string[];
  model_prefixes: string[];
  timeout_ms: number;
  max_tokens_field: string;
  secret_ref: string | null;
  has_api_key: boolean;
  health_status: string;
  created_at: number;
  updated_at: number;
}

export interface GatewayLog {
  id: number;
  time: number;
  client_type: string;
  endpoint: string;
  requested_model: string;
  resolved_provider: string | null;
  resolved_model: string | null;
  protocol_in: string;
  protocol_out: string | null;
  status_code: number;
  latency_ms: number;
  stream: boolean;
  token_estimate: number | null;
  input_tokens: number | null;
  output_tokens: number | null;
  total_tokens: number | null;
  cache_creation_input_tokens: number | null;
  cache_read_input_tokens: number | null;
  fidelity_mode: string;
  error_message: string | null;
}

export const gatewayApi = {
  getStatus: () => invoke<GatewayStatus>("get_gateway_status"),
  start: () => invoke<GatewayStatus>("start_gateway"),
  stop: () => invoke<void>("stop_gateway"),
  restart: () => invoke<GatewayStatus>("restart_gateway"),
  getApiKeyStatus: () => invoke<ApiKeyStatus>("get_local_api_key_status"),
  regenerateApiKey: () => invoke<string>("regenerate_local_api_key"),
  getApiKeyOnce: () => invoke<string | null>("get_local_api_key_once"),
  getProviders: () => invoke<Provider[]>("get_providers"),
  getLogs: (limit = 100) => invoke<GatewayLog[]>("get_gateway_logs", { limit }),
  clearLogs: () => invoke<void>("clear_gateway_logs"),
};

File: src-tauri/src/commands/gateway.rs
use crate::models::{ApiKeyStatus, GatewayLog, GatewayStatus, Provider};
use crate::store::AppState;
use tauri::State;

#[tauri::command]
pub fn get_gateway_status(state: State<'_, AppState>) -> Result<GatewayStatus, String> {
    state.gateway.status(&state.db).map_err(Into::into)
}

#[tauri::command]
pub fn start_gateway(state: State<'_, AppState>) -> Result<GatewayStatus, String> {
    state.gateway.start(state.db.clone()).map_err(Into::into)
}

#[tauri::command]
pub fn stop_gateway(state: State<'_, AppState>) -> Result<(), String> {
    state.gateway.stop().map_err(Into::into)
}

#[tauri::command]
pub fn restart_gateway(state: State<'_, AppState>) -> Result<GatewayStatus, String> {
    state.gateway.restart(state.db.clone()).map_err(Into::into)
}

#[tauri::command]
pub fn get_local_api_key_status(state: State<'_, AppState>) -> Result<ApiKeyStatus, String> {
    state.db.get_local_api_key_status().map_err(Into::into)
}

#[tauri::command]
pub fn regenerate_local_api_key(state: State<'_, AppState>) -> Result<String, String> {
    state.db.regenerate_local_api_key().map_err(Into::into)
}

#[tauri::command]
pub fn get_local_api_key_once(state: State<'_, AppState>) -> Result<Option<String>, String> {
    state.db.get_local_api_key_once().map_err(Into::into)
}

#[tauri::command]
pub fn get_providers(state: State<'_, AppState>) -> Result<Vec<Provider>, String> {
    state.db.get_providers().map_err(Into::into)
}

#[tauri::command]
pub fn get_gateway_logs(
    state: State<'_, AppState>,
    limit: Option<i64>,
) -> Result<Vec<GatewayLog>, String> {
    state
        .db
        .get_gateway_logs(limit.unwrap_or(100))
        .map_err(Into::into)
}

#[tauri::command]
pub fn clear_gateway_logs(state: State<'_, AppState>) -> Result<(), String> {
    state.db.clear_gateway_logs().map_err(Into::into)
}
